### PR TITLE
ARROW-9078: [C++] Parquet read / write extension type with nested storage type

### PR DESCRIPTION
**Pull request by pitrou**

### CODE REVIEW

File: cpp/src/arrow/extension_type.h [hunk 1]
```diff
@@ -80,6 +81,12 @@ class ARROW_EXPORT ExtensionType : public DataType {
   /// \return the serialized representation
   virtual std::string Serialize() const = 0;
 
+  static std::shared_ptr<Array> WrapArray(const std::shared_ptr<DataType>& ext_type,
+                                          const std::shared_ptr<Array>& storage);
+  static std::shared_ptr<ChunkedArray> WrapArray(
+      const std::shared_ptr<DataType>& ext_type,
+      const std::shared_ptr<ChunkedArray>& storage);
```
Comment (pitrou) [inline cpp/src/arrow/extension_type.h:88]: @wesm Thoughts about this API?

Comment (wesm) [reply cpp/src/arrow/extension_type.h:88]: Seems OK to me. May want doxygen comments

Comment (pitrou) [reply cpp/src/arrow/extension_type.h:88]: Right :-)

**Comment by pitrou** [comment]
@emkornfield Would you like to take a quick look?

File: cpp/src/parquet/arrow/schema.cc [hunk 2]
```diff
@@ -682,48 +689,78 @@ Status GetOriginSchema(const std::shared_ptr<const KeyValueMetadata>& metadata,
...
+
+Status ApplyOriginalMetadata(const Field& origin_field, SchemaField* inferred) {
+  auto origin_type = origin_field.type();
+  auto inferred_type = inferred->field->type();
+
+  if (origin_type->id() == ::arrow::Type::EXTENSION) {
+    const auto& ex_type = checked_cast<const ::arrow::ExtensionType&>(*origin_type);
+    auto origin_storage_field = origin_field.WithType(ex_type.storage_type());
+
+    // Apply metadata recursively to storage type
```
Comment (emkornfield) [inline cpp/src/parquet/arrow/schema.cc:751]: this seems to conflict with the TODO above.

File: cpp/src/parquet/arrow/schema.cc [hunk 3]
```diff
@@ -682,48 +689,78 @@ Status GetOriginSchema(const std::shared_ptr<const KeyValueMetadata>& metadata,
...
+    RETURN_NOT_OK(ApplyOriginalStorageMetadata(*origin_storage_field, inferred));
+    inferred->storage_field = inferred->field;
+
+    // Restore extension type, if the storage type is as read from Parquet
```
Comment (emkornfield) [inline cpp/src/parquet/arrow/schema.cc:755]: isn't everything here read from Parquet (is this in contrast to user provided?)

File: cpp/src/parquet/arrow/writer.cc [hunk 4]
```diff
@@ -63,6 +66,10 @@ using arrow::Status;
 using arrow::Table;
 using arrow::TimeUnit;
 
+using ArrowTypeId = arrow::Type;
```
Comment (emkornfield) [inline cpp/src/parquet/arrow/writer.cc:69]: nit: this actually makes the code less understandable to me and it seems like it only saves 1 character?  (not a big deal either way though).

**Comment by emkornfield** [comment]
A few minor doc comments otherwise looks ok to me.

**Review by jorisvandenbossche — COMMENTED** [review]
Cool, thanks a lot for looking into this! 
(only checked the python tests for now, which looks good)

I tested it with a geometry extension type I am working on (using `list<parts: list<rings: list<vertices: list<xy: double not null> not null> not null>>` as the storage type), and that seems to work nicely.

One gotcha I ran into while using a fixed size list type (nested within another list) that in this case the extension type wasn't preserved. But I suppose this is because the fixed size list is not yet preserved on parquet roundtrip (read back as normal list type), and to add back the extension type, the storage type needs to be the same. So in hindsight, this behaviour is as expected (as long as fixed size list is not yet supported).

**Comment by emkornfield** [comment]
> One gotcha I ran into while using a fixed size list type (nested within another list) that in this case the extension type wasn't preserved. But I suppose this is because the fixed size list is not yet preserved on parquet roundtrip (read back as normal list type), and to add back the extension type, the storage type needs to be the same. So in hindsight, this behaviour is as expected (as long as fixed size list is not yet supported).

@jorisvandenbossche Reading fixed size lists is not yet supported.  One use-case that will likely not be supported in the short term is reading parquet files with FixedSizeLists that have null elements.  Do your use-cases with FixedSizeList allow for null lists?

**Comment by jorisvandenbossche** [comment]
@emkornfield for my *personal* use case (in context of https://github.com/geopandas/geo-arrow-spec/, see this notebooks with some proposed nested types: https://nbviewer.jupyter.org/gist/jorisvandenbossche/00e5c4a54f7b94375ccc6921c07825a0), the inner fixed size list would not allow nulls, only the most outer variable size list would be nullable.

File: cpp/src/parquet/arrow/schema.cc (hunk 3 above)
Comment (pitrou) [reply cpp/src/parquet/arrow/schema.cc:755]: Hmm... right. Would "as inferred from Parquet" be clearer?

File: cpp/src/parquet/arrow/writer.cc (hunk 4 above)
Comment (pitrou) [reply cpp/src/parquet/arrow/writer.cc:69]: I can revert :-)

File: cpp/src/parquet/arrow/schema.cc (hunk 2 above)
Comment (pitrou) [reply cpp/src/parquet/arrow/schema.cc:751]: I should make the TODO more precise then, it is about applying to children types.

**Comment by pitrou** [comment]
I rebased and addressed review comments.

**Comment by pitrou** [comment]
Will merge.

**Pull request closed without merging**